exchange/exchange.go

```diff
@@ -345,57 +356,86 @@ func applyCategoryMapping(requestExt openrtb_ext.ExtRequest, seatBids *map[openr
 	for bidderName, seatBid := range *seatBids {
 		bidsToRemove := make([]int, 0)
 		for bidInd := range seatBid.bids {
-			bid := seatBid.bids[bidInd].bid
+			bid := seatBid.bids[bidInd]
 			var duration int
 			var category string
-			var tempRespBidExt openrtb_ext.ExtBid
-			if len(bid.Ext) > 0 {
-				if err := json.Unmarshal(bid.Ext, &tempRespBidExt); err != nil {
-					return res, err
-				}
+			var pb string
 
-				//handler for different extension formats
-				var adapterExt json.RawMessage
-				if tempRespBidExt.Bidder != nil {
-					adapterExt = tempRespBidExt.Bidder
+			if bid.bidVideo != nil {
+				duration = bid.bidVideo.Duration
+				category = bid.bidVideo.PrimaryCategory
+			}
+
+			if category == "" {
+				bidIabCat := bid.bid.Cat
+				if len(bidIabCat) != 1 {
+					//TODO: add metrics
+					//on receiving bids from adapters if no unique IAB category is returned  or if no ad server category is returned discard the bid
+					bidsToRemove = append(bidsToRemove, bidInd)
+					continue
 				} else {
-					adapterExt = bid.Ext
+					//if unique IAB category is present then translate it to the adserver category based on mapping file
+					category, err = categoriesFetcher.FetchCategories(primaryAdServer, publisher, bidIabCat[0])
+					if err != nil || category == "" {
+						//TODO: add metrics
+						//if mapping required but no mapping file is found then discard the bid
+						bidsToRemove = append(bidsToRemove, bidInd)
+						continue
+					}
 				}
+			}
+
+			// TODO: consider should we remove bids with zero duration here?
+
+			pb, err = GetCpmStringValue(bid.bid.Price, targData.priceGranularity)
```

Comment:
I see GetCpmStringValue never returns an error.  Maybe should replace _err_ above with an underscore since it doesn't seem necessary to check err.

Reply:
OTOH, should there be checks of _pb_ to make sure it's at least not an empty string?  Seems like an empty string value is a possibility.

Reply:
@josephveach Thanks for spotting. 
We will definitely change err to underscore as in no case we would get an err. However since this function was a pre-existing function we would like to defer to @hhhjort to specify the behavior for empty pb. Imo if no price then the bid seems to be invalid. I would recommend to address this in a separate PR as this behavior should have always been present even prior  to this change.

Reply:
I do recall some discussions on if a zero bid bid is valid or not. I don't recall what the final result was off the top of my head. The function should return a non blank `pb` as long as a valid price granularity exists. I am not sure how tight the validation on granularities is, so it might be possible.